### PR TITLE
Add interstitial page review actions when creating new stages.

### PR DESCRIPTION
**Pull request by dragonfax**
### Screenshot

samson

### References

* JIRA - https://zendesk.atlassian.net/browse/TOOLS-872

### Risks

None

Just adding a read-only page to an existing workflow.

### CODE REVIEW

File: app/controllers/admin/deploy_groups_controller.rb [hunk 1]
```diff
@@ -2,8 +2,9 @@
 class Admin::DeployGroupsController < ApplicationController
   before_action :authorize_admin!
   before_action :authorize_super_admin!, only: [:create, :new, :update, :destroy, :deploy_all,
-                                                :create_all_stages, :edit]
-  before_action :deploy_group, only: [:show, :edit, :update, :destroy, :deploy_all, :create_all_stages]
+                                                :create_all_stages, :create_all_stages_verify, :edit]
+  before_action :deploy_group, only: [:show, :edit, :update, :destroy, :deploy_all, :create_all_stages,
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]:
```Ruby
  before_action :authorize_super_admin!, only: [
    :create, :new, :update, :destroy, :deploy_all, :create_all_stages, :create_all_stages_verify, :edit
  ]
  before_action :deploy_group, only: [
    :show, :edit, :update, :destroy, :deploy_all, :create_all_stages, :create_all_stages_verify
  ]
```

File: app/controllers/admin/deploy_groups_controller.rb [hunk 2]
```diff
@@ -74,23 +75,48 @@ def deploy_all
     end
   end
 
+  def create_all_stages_verify
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: `create_all_stages_preview` ?

File: app/controllers/admin/deploy_groups_controller.rb [hunk 3]
```diff
@@ -74,23 +75,48 @@ def deploy_all
...
+    @stages_preexisting, @stages_to_create = gather_stages_for_creation
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: gather is kind of meaningless ... `stages_for_creation `

File: app/controllers/admin/deploy_groups_controller.rb [hunk 4]
```diff
@@ -74,23 +75,48 @@ def deploy_all
...
+
+    render
+  end
+
   def create_all_stages
     # No more than one stage, per project, per deploy_group
     # Note: you can call this multiple times, and it will create missing stages, but no redundant stages.
+
+    _throwaway, stages_to_create = gather_stages_for_creation
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: `_throwaway` is meaningless ... 

prefer `_, stages = foo` or `stages = foo.last`

File: app/controllers/admin/deploy_groups_controller.rb [hunk 5]
```diff
@@ -74,23 +75,48 @@ def deploy_all
...
+    stages_to_create.each do |template_stage|
+      create_stage_with_group(template_stage)
+    end
+
+    redirect_to [:admin, deploy_group]
+  end
+
+  private
+
+  # gather_stages_for_creation()
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: can remove this line ... does not add any value :D

File: app/controllers/admin/deploy_groups_controller.rb [hunk 6]
```diff
@@ -74,23 +75,48 @@ def deploy_all
     end
   end
 
+  def create_all_stages_verify
+    @stages_preexisting, @stages_to_create = gather_stages_for_creation
+
+    render
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: no need to call that

File: app/controllers/admin/deploy_groups_controller.rb [hunk 7]
```diff
@@ -74,23 +75,48 @@ def deploy_all
...
   def create_all_stages
     # No more than one stage, per project, per deploy_group
     # Note: you can call this multiple times, and it will create missing stages, but no redundant stages.
+
+    _throwaway, stages_to_create = gather_stages_for_creation
+    stages_to_create.each do |template_stage|
+      create_stage_with_group(template_stage)
+    end
+
+    redirect_to [:admin, deploy_group]
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: `notice: "15 stages created"` or so ?

File: app/controllers/admin/deploy_groups_controller.rb [hunk 8]
```diff
@@ -74,23 +75,48 @@ def deploy_all
...
+        stages_preexisting << deploy_group_stage
+      elsif template_stage
+        stages_to_create << template_stage
       end
     end
-    redirect_to [:admin, deploy_group]
-  end
 
-  private
+    [stages_preexisting, stages_to_create]
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: stages_preexisting sounds kind of clunky ... preexisting_stages ? ... does break the symmetry sadly ...

Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: maybe `preexisting` and `missing`

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 9]
```diff
@@ -0,0 +1,79 @@
+<h1>Deploy to <%= @deploy_group.name %></h1>
+
+<label>Environment</label>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: would make more sense as a single `<h1>` like `Creating stages for <name>` ?

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+<small>These stages will be created.</small>
+
+<section class="clearfix">
+  <div class="table table-hover table-condensed">
+    <table class="table">
+      <thead>
+        <tr>
+          <th>Project</th>
+          <th>New Stage Name</th>
+          <th>Copied From</th>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: `From template` to reiterate where the info is coming from ?

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        </tr>
+      </thead>
+
+      <tbody>
+        <% @stages_to_create.each do |template_stage| %>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: if these are template_stages then the variable should be stages_to_copy ...

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+  </div>
+
+  <div class="admin-actions">
+    <div class="pull-right">
+      <%= link_to "Create Stages",
+        create_all_stages_admin_deploy_group_path(@deploy_group),
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: maybe indent more to make it visually be inside of the link_to ... wrapping paranthesese could help too

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        class: "btn btn-primary",
+        data: {
+          method: "post",
+          confirm: "All Project marked 'Include in New Deploy Groups' will get a stage for this deploy group.\n\n" +
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: this text is not on this page ... maybe remove and make the button `Creation 15 stages`

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        }
+      %>
+
+      <%= link_to "Cancel", admin_deploy_group_path(@deploy_group), class: "btn btn-default"%>
+    </div>
+  </div>
+</section>
+
+
+<h2>Stages Already Created</h2>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: simplify these 2 lines into `Existing stages for <deploy group>` ?

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+          <th>Project</th>
+          <th>Stage Name</th>
+        </tr>
+      </thead>
+
+      <tbody>
+        <% @stages_preexisting.each do |stage| %>
+          <tr>
+            <td><%= stage.project.name %></td>
+            <td><%= stage.name %></td>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: maybe list deploy groups here, they would show pod1/pod2 etc might resolve some confusion and would make it symmetric to the upper table

File: app/views/admin/deploy_groups/create_all_stages_verify.html.erb [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+  </div>
+</section>
+
+
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_verify.html.erb]: remove whitespace

File: app/views/admin/deploy_groups/show.html.erb [hunk 17]
```diff
@@ -36,12 +36,8 @@
       <%= link_to "Edit", edit_admin_deploy_group_path(@deploy_group), class: "btn btn-primary" %>
       <% if DeployGroup.enabled? %>
         <%= link_to "Create Stages",
```
Comment (grosser) [inline app/views/admin/deploy_groups/show.html.erb]: misleading now ... would be better as `Create Stages ...` or some other kind of indicator that it's not dangerous

**Comment by dragonfax** [comment]
Your using github's code review flow? Have we accepted that as a standard at Zendesk yet? I haven't seen any emails about it.

I have problems with githubs Code Review system. Its too strict and locks the reviews down to you specifically, one reviewer. For instance, I didn't even submit this code for review. I was just getting a CI run on it. I never mentioned anyone.

**Comment by grosser** [comment]
Review was OK-ed by compliance ... there was a email thread on dev@

Just saw the PR and jumped on it ;)

**Comment by grosser** [comment]
review system does not lock anything down to a user afaik ... anyone can leave a review

**Comment by dragonfax** [comment]
Don't get the wrong idea, your comments are welcome. And I'll address all of them.

**Comment by dragonfax** [comment]
But only the one who left a review, can approve the code after that point.

**Comment by grosser** [comment]
works ... `@craig-day` said that was a bug they fixed

**Comment by dragonfax** [comment]
@grosser Updated the screenshot too.

File: app/controllers/admin/deploy_groups_controller.rb [hunk 18]
```diff
@@ -74,23 +75,44 @@ def deploy_all
     end
   end
 
+  def create_all_stages_preview
+    @preexisting_stages, @missing_stages = stages_for_creation
+  end
+
   def create_all_stages
     # No more than one stage, per project, per deploy_group
```
Comment (grosser) [inline app/controllers/admin/deploy_groups_controller.rb]: usually method level comments go on the outside ...

File: app/views/admin/deploy_groups/create_all_stages_preview.html.erb [hunk 19]
```diff
@@ -0,0 +1,78 @@
+<h1>Creating Stages for <%= @deploy_group.name %> in <%= @deploy_group.environment.name %></h1>
+
+<h2>Creating Stages</h2>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_preview.html.erb]: redundant ?

File: app/views/admin/deploy_groups/create_all_stages_preview.html.erb [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+            <th>Deploy Groups</th>
+          </tr>
+        </thead>
+
+        <tbody>
+          <% @preexisting_stages.each do |stage| %>
+            <tr>
+              <td><%= stage.project.name %></td>
+              <td><%= stage.name %></td>
+              <td><%= stage.deploy_groups.map(&:name).join(', ') %></td>
```
Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_preview.html.erb]: `.sort_by(&:natural_order)`

File: app/views/admin/deploy_groups/create_all_stages_preview.html.erb (hunk 19 above)
Comment (dragonfax) [inline app/views/admin/deploy_groups/create_all_stages_preview.html.erb]: Looks a little different than the h1, and works to clarify the 2 tables seperately. See the screenshot.

Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_preview.html.erb]: Maybe "New staged" ... or "To be created"

Comment (grosser) [inline app/views/admin/deploy_groups/create_all_stages_preview.html.erb]: ... but does not really matter ...

**Pull request merged**